autoprecompiles/src/execution/evaluator.rs

```diff
@@ -0,0 +1,471 @@
+use std::{collections::HashMap, sync::Arc};
+
+use itertools::Itertools;
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    execution::{
+        ast::{
+            LocalOptimisticLiteral, OptimisticConstraint, OptimisticExpression, OptimisticLiteral,
+        },
+        ExecutionState,
+    },
+    powdr::UniqueReferences,
+};
+
+/// A collection of optimistic constraints over the intermediate execution states of a block, to be accessed in chronological order
+#[derive(Debug, Serialize, Deserialize, deepsize2::DeepSizeOf)]
+pub struct OptimisticConstraints<A, V> {
+    /// For each step, the execution values we need to remember for future constraints, excluding this step
+    fetches_by_step: Vec<(usize, Vec<LocalOptimisticLiteral<A>>)>,
+    /// For each step, the constraints that must be satisfied
+    constraints_to_check_by_step: Vec<(usize, Vec<OptimisticConstraint<A, V>>)>,
```

Comment:
```suggestion
    /// For instruction index, the execution values (if any) we need to remember for future constraints, excluding this step
    fetches_by_step: HashMap<usize, Vec<LocalOptimisticLiteral<A>>>,
    /// For instruction index, the constraints (if any) that must be satisfied
    constraints_to_check_by_step: HashMap<usize, Vec<OptimisticConstraint<A, V>>>,
```

Would it make sense to directly map from instruction index, with no entry if the list is empty? I find this clearer, and you wouldn't need to maintain a `constraints_index` and `fetches_index` in `OptimisticConstraintEvaluator`.

The only argument against this I see is maybe performance, is that the motivation?

Reply:
Oh I think I just made a similar comment ;)

If performance, it's probably cuz with Vec we loop over instructions with fetches and instructions with constraint checks only, instead of calling get on every single instruction index.

Concretely can be faster though seem asymptotically the same.